src/main/scala/scala/tools/refactoring/implementations/MarkOccurrences.scala

```diff
@@ -5,67 +5,229 @@
 
-abstract class MarkOccurrences extends common.Selections with analysis.Indexes with common.CompilerAccess with common.EnrichedTrees {
-
-  val global: tools.nsc.interactive.Global
+trait MarkOccurrences extends common.Selections with analysis.Indexes with common.CompilerAccess with common.EnrichedTrees with common.InteractiveScalaCompiler {
   import global._
 
-  def occurrencesOf(file: tools.nsc.io.AbstractFile, from: Int, to: Int): (Tree, List[Position]) = {
+  protected class SingleTreeSelection(val selected: Tree, val root: Tree) {
+    val symbol = selected match {
+      case Import(expr, List(selector)) =>
+        findSymbolForImportSelector(expr, selector.name).getOrElse(NoSymbol)
+      case tree => tree.symbol
+    }
 
-    def positionOverlapsSelection(p: Position) = (
-          ((new RangePosition(null, from, from, to)) overlaps p) ||
-          (from == to && ( to == p.end || to == p.start))
-      )
+    val name = {
+      // We try to read the old name from the name position of the original symbol, since this seems to be the most
+      // reliable strategy that also works well in the presence of `backtick-identifiers`.
+      index.declaration(symbol).flatMap { tree =>
+        tree.namePosition() match {
+          case rp: RangePosition => Some(rp.source.content.slice(rp.start, rp.end).mkString(""))
+          case _ => None
+        }
+      }.getOrElse {
+        trace("Cannot find old name of symbol reliably; attempting fallback")
+        try {
+          selected match {
+            case Import(_, List(selector)) => selector.rename.toString
+            case _ => selected match {
+              case s: Select => s.name.decoded
+              case other => other.nameString
+            }
+          }
+        } catch {
+          case NonFatal(e) =>
+            trace("Cannot find old name of symbol; giving up")
+            "<no-name>"
+        }
+      }
+    } \\ { name =>
+      trace(s"SingleTreeSelection.name: $name")
+    }
+  }
 
-    def occurrencesForSymbol(s: Symbol) = {
-      val occurrences = index.occurences(s)
-      occurrences map (_.namePosition) filter  (_.isRange)
+  protected def findOccurrences(selection: SingleTreeSelection): List[(RangePosition, Tree)] = {
+    /*
+     * A lazy val is represented by a ValDef and an associated DefDef that contains the initialization code.
+     * Unfortunately, the Scala compiler does not set modifier positions for the DefDef, which is a problem for us,
+     * because we are relying on the DefDef when printing out source code. The following function patches the affected
+     * DefDefs accordingly.
+     *
+     * See #1002392 and #1002502
+     */
+    def eventuallyFixModifierPositionsForLazyVals(t: Tree): Unit = t match {
+      case dd: DefDef if dd.mods.isLazy && dd.mods.positions.isEmpty =>
+        val vd = selection.root.find {
+          case vd: ValDef if vd.mods.isLazy && !vd.mods.positions.isEmpty && dd.pos.point == vd.pos.point => true
```

Comment:
To point is used to make sure that we are indeed selecting the right `ValDef` to patch.